.Net: Support for configuring dimensions in Google AI embeddings generation

ArieSLV opened the pull request:
### Motivation and Context

This change addresses a limitation in the current implementation of the Google AI embeddings generation service in Semantic Kernel. Currently, users cannot configure the output dimensionality of the embeddings, even though the underlying Google AI API supports specifying the number of dimensions via the `output_dimensionality` parameter.

**Why is this change required?**  
Allowing configuration of the dimensions provides greater flexibility for users to tailor the embeddings to their specific use cases—whether for optimizing memory usage, improving performance, or ensuring compatibility with downstream systems that expect a particular embedding size.

**What problem does it solve?**  
It solves the issue of inflexibility by exposing the `dimensions` parameter in the service constructors, builder methods, and API request payloads. This ensures that developers can leverage the full capabilities of the Google API without being limited to the default embedding size.

**What scenario does it contribute to?**  
This feature is particularly useful in scenarios where:
- Users need to optimize storage or computational resources.
- Downstream tasks or integrations require embeddings of a specific dimensionality.
- Fine-tuning the model output is essential for performance or compatibility reasons.

Relevant issue link: https://github.com/microsoft/semantic-kernel/issues/10488

### Description

This PR introduces support for specifying the output dimensionality in the Google AI embeddings generation workflow. The main changes include:

- **Service Constructor Update:**  
  The `GoogleAITextEmbeddingGenerationService` constructor now accepts an optional `dimensions` parameter, which is then forwarded to the lower-level client implementations.

- **Builder and Extension Methods:**  
  Extension methods such as `AddGoogleAIEmbeddingGeneration` have been updated to accept a `dimensions` parameter. This allows developers to configure the embedding dimensions using the builder pattern.

- **Request Payload Enhancement:**  
  The `GoogleAIEmbeddingRequest` class now includes a new optional property `Dimensions` (serialized as `output_dimensionality`). When provided, this value is included in the JSON payload sent to the Google AI API.

- **Metadata and Attributes Update:**  
  The service’s metadata now reflects the provided dimensions, ensuring consistency in configuration tracking.

- **Unit Testing:**  
  New unit tests have been added to confirm that:
  - When a `dimensions` value is provided, it is correctly included in the JSON request.
  - When not provided, the default behavior remains unchanged.
  
This enhancement maintains backward compatibility since the new parameter is optional. Existing implementations that do not specify a dimension will continue to work as before.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings.
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations.
- [x] All unit tests pass, and I have added new tests where possible.
- [x] I didn't break anyone :smile:

ArieSLV commented:
Hi team,
I see the Contributor License Agreement (CLA) message and understand that I need to agree to it before my submission can be processed. I've reached out to my manager for approval to sign the CLA under my employer’s name. Once I receive the necessary clearance, I will reply accordingly.
I appreciate your patience.

ArieSLV commented:
@microsoft-github-policy-service agree

ArieSLV commented:
Hi team,
The Contributor License Agreement issue has been resolved. Are any further actions required from me to get a review?

RogerBarreto commented on dotnet/src/Connectors/Connectors.Google.UnitTests/Services/GoogleAITextEmbeddingGenerationServiceTests.cs:
```diff
@@ -8,14 +8,63 @@ namespace SemanticKernel.Connectors.Google.UnitTests.Services;
         // Assert
-        Assert.Equal(model, service.Attributes[AIServiceExtensions.ModelIdKey]);
+        Assert.Null(result);
     }
```
Please also add tests checking that the `RequestContent` generated the right `property` in the body before calling the API.

like:

1. `Dimensions = null` don't find `output_dimensionality` property in the request
2. `Dimensions = 2` should find `output_dimensionality` property in the request with value  = 2.

Similar to this:

https://github.com/microsoft/semantic-kernel/blob/4fdaf67b5a6314311676bea2a36a3b3dc2c08662/dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Services/OpenAITextToImageServiceTests.cs#L120

ArieSLV replied:
I've added the requested tests in the latest changes:

- Added `ShouldNotIncludeDimensionsInRequestWhenNotProvidedAsync()` which verifies no `output_dimensionality` property is sent when dimensions = null
- Added `ShouldIncludeDimensionsInRequestWhenProvidedAsync()` which checks that the `output_dimensionality` property is included with the correct value when dimensions are specified

The tests use a similar pattern to the OpenAI tests you referenced, with HTTP message handler stubbing to capture and verify the actual request content being sent to the API.

RogerBarreto commented:
Some small spell check errors to fix.

ArieSLV commented on dotnet/src/SemanticKernel.Abstractions/Services/AIServiceExtensions.cs:
```diff
@@ -49,6 +54,17 @@ public static class AIServiceExtensions
+    /// </summary>
+    /// <param name="service">The service from which to get the dimensions.</param>
+    /// <returns>The dimensions if it was specified in the service's attributes; otherwise, null.</returns>
+    public static int? GetDimensions(this IAIService service)
```
Thanks Roger. I've moved the `DimensionsKey` constant and `GetDimensions()` method from `AIServiceExtensions` to `EmbeddingGenerationExtensions` as suggested.

ArieSLV commented:
Hi @RogerBarreto,

Thanks for the thorough review. I've addressed all your feedback in the latest changes:

- Added integration tests in `EmbeddingGenerationTests.cs` with both default and custom dimensions scenarios
- Created a sample implementation in `Concepts\Memory\Google_EmbeddingGeneration.cs` showing how to use the new functionality
- Added unit tests that verify proper request formatting with and without dimensions

Regarding the spell check errors, I believe those might have been resolved during the merge as they weren't directly related to my code changes.

markwallace-microsoft commented on dotnet/src/Connectors/Connectors.Google/Core/GoogleAI/GoogleAIEmbeddingRequest.cs:
```diff
@@ -45,5 +46,9 @@ internal sealed class RequestEmbeddingContent
+
+        [JsonPropertyName("output_dimensionality")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public int? Dimensions { get; set; }
```
Should the name be `OutputDimensionality`?

Also there is inconsistent naming for the JSON properties, here it is correct about for `taskType` it should be `task_type`.
@RogerBarreto perhaps we can fix that in a seperate PR, thoughts?

RogerBarreto replied:
by looking at the documentation `output_dimensionality` actually can be `outputDimensionality`, it's not cobra_cased, funily enouth I did run the integration tests and the dimentions worked with cobra_cased.

Reference:
https://ai.google.dev/api/embeddings#EmbedContentRequest

@ArieSLV Can you have a look on this please?
